fix(aio): activate ServiceWorker updates asap

gkalpak opened the pull request:
Previouly, whenever a new ServiceWorker update was detected the user was prompted to update (with a notification). This turned out to be more distracting than helpful. Also, one would get notifications on all open browser tabs/windows and had to manually reload each one in order for the whole content (including the app) to be updated.

This commit changes the update strategy as follows:
- Whenever a new update is detected, it is immediately activated (and all tabs/windows will be notified).
- Once an update is activated (regardless of whether the activation was initiated by the current tab/window or not), a flag will be set to do a "full page navigation" the next time the user navigates to a document.

Benefits:
- All tabs/windows are updated asap.
- The updates are applied authomatically, without the user's needing to do anything.
- The updates are applied in a way that:
  a. Ensures that the app and content versions are always compatible.
  b. Does not distract the user from their usual workflow.

NOTE:
The "full page navigation" may cause a flash (while the page is loading from scratch), but this is expected to be minimal, since at that point almost all necessary resources are cached by and served from the ServiceWorker.

Fixes #17539.

gkalpak committed:
fix(aio): activate ServiceWorker updates asap

Previouly, whenever a new ServiceWorker update was detected the user was
prompted to update (with a notification). This turned out to be more distracting
than helpful. Also, one would get notifications on all open browser tabs/windows
and had to manually reload each one in order for the whole content (including
the app) to be updated.

This commit changes the update strategy as follows:
- Whenever a new update is detected, it is immediately activated (and all
  tabs/windows will be notified).
- Once an update is activated (regardless of whether the activation was
  initiated by the current tab/window or not), a flag will be set to do a
  "full page navigation" the next time the user navigates to a document.

Benefits:
- All tabs/windows are updated asap.
- The updates are applied authomatically, without the user's needing to do
  anything.
- The updates are applied in a way that:
  a. Ensures that the app and content versions are always compatible.
  b. Does not distract the user from their usual workflow.

NOTE:
The "full page navigation" may cause a flash (while the page is loading from
scratch), but this is expected to be minimal, since at that point almost all
necessary resources are cached by and served from the ServiceWorker.

Fixes #17539

mary-poppins commented:
The angular.io preview for ced52d6fa5d2b6787062deeb6997099858b2e74e is available [here][1].

[1]: https://pr17699-ced52d6fa5d2b6787062deeb6997099858b2e74e.ngbuilds.io/

mary-poppins commented:
You can preview 1b866549ef850b99a26196563d0cbb12281d7454 at https://pr17699-1b866549ef850b99a26196563d0cbb12281d7454.ngbuilds.io/.

mary-poppins commented:
You can preview 311c78322dada1e41c13274e96ca374203f2a556 at https://pr17699-311c78322dada1e41c13274e96ca374203f2a556.ngbuilds.io/.

mary-poppins commented:
You can preview 9cc698702087269d8ad9a04a88f993b120c2794f at https://pr17699-9cc698702087269d8ad9a04a88f993b120c2794f.ngbuilds.io/.

mary-poppins commented:
You can preview 2af11645419862f11f829b3be53cf0e2a93ffb49 at https://pr17699-2af11645419862f11f829b3be53cf0e2a93ffb49.ngbuilds.io/.

mary-poppins commented:
You can preview 368aabc9490a50b87b72201a1d6568b2cfaa4b1e at https://pr17699-368aabc9490a50b87b72201a1d6568b2cfaa4b1e.ngbuilds.io/.

petebacondarwin commented on aio/src/app/sw-updates/sw-updates.module.ts:
```diff
@@ -1,20 +1,14 @@
 import { NgModule } from '@angular/core';
-import { MdSnackBarModule } from '@angular/material';
```
Does this removal help lower our bundle sizes or is the snack bar used elsewhere?

gkalpak replied:
No, the copy button uses it.

petebacondarwin commented on aio/src/app/sw-updates/sw-updates.service.spec.ts:
```diff
@@ -14,11 +15,13 @@ describe('SwUpdatesService', () => {
+  //   Because `SwUpdatesService` uses the `debounceTime` operator, it needs to be instantiated
+  //   inside the `fakeAsync` zone (when `fakeAsync` is used for the test). Thus, we can't run
+  //   `setup()` in a `beforeEach()` block. We use the `run()` helper to call `setup()` inside each
+  //   test's zone.
```
Why is it the case that we cannot use a `beforeEach`? Surely we can just use the `fakeAsync` helper there too?

```ts
beforeEach(fakeAsync() => {
  injector = ReflectiveInjector.resolveAndCreate([
    { provide: Logger, useClass: MockLogger },
    { provide: NgServiceWorker, useClass: MockNgServiceWorker },
    SwUpdatesService
  ]);
});
```

gkalpak replied:
Not all `it`s use `fakeAsync`.

petebacondarwin replied:
Does that matter?

gkalpak replied:
IIRC, if the `beforeEach` is in `fakeAsync` then everything has to be and if everything is in `fakeAsync` then it complains about pending tasks. It is solveable (by flushing the tasks or something), but back when I first wrote the tests it seemed better to do it like this.

(I don't feel strongly about changing it, if you think it is better in an other way.)

petebacondarwin commented on aio/src/app/sw-updates/sw-updates.service.spec.ts:
```diff
@@ -51,163 +54,103 @@ describe('SwUpdatesService', () => {
   })));
 
-  it('should not schedule a new check if there is an update available', fakeAsync(run(() => {
+  it('should activate new updates immediatelly', fakeAsync(run(() => {
```
typo `immediatelly`

petebacondarwin commented on aio/src/app/sw-updates/sw-updates.service.ts:
```diff
@@ -17,60 +18,58 @@ import 'rxjs/add/operator/toPromise';
 @Injectable()
 export class SwUpdatesService implements OnDestroy {
-  private checkInterval = 1000 * 60 * 60 * 6;   // 6 hours
+  private checkInterval = 1000 /* 60 * 60 */* 6;   // 6 hours
```
Ahem!

petebacondarwin commented:
Why use debounce in the `SwUpdatesService` rather than a more simple `setInterval`?

petebacondarwin commented:
A rebase should fix Travis now we have landed the lighthouse fix

gkalpak commented:
> Why use debounce in the SwUpdatesService rather than a more simple setInterval?

Because everything is an observable :grin:

mary-poppins commented:
You can preview a1129c7 at https://pr17699-a1129c7.ngbuilds.io/.

petebacondarwin commented:
I was thinking of something along the lines of

```typescript
@Injectable()
export class SwUpdatesService implements OnDestroy {
  private checkInterval = 1000 * 60 * 60 * 6;   // 6 hours
  private onDestroy = new Subject();
  private checkForUpdateSubj = new Subject();
  updateActivated = this.sw.updates.
                      takeUntil(this.onDestroy)
                      .do(evt => this.log(`Update event: ${JSON.stringify(evt)}`))
                      .filter(evt => evt.type === 'activation')
                      .map(evt => evt.version)
                      .startWith(null);

  constructor(private logger: Logger, private sw: NgServiceWorker) {
    Observable.timer(0, this.checkInterval)
      .takeUntil(this.onDestroy)
      .subscribe(() => this.checkForUpdate());
  }

  ngOnDestroy() {
    this.onDestroy.next();
  }

  private activateUpdate() {
    this.log('Activating update...');
    this.sw.activateUpdate(null).subscribe();
  }

  private checkForUpdate() {
    this.log('Checking for update...');
    this.sw.checkForUpdate()
        // Temp workaround for https://github.com/angular/mobile-toolkit/pull/137.
        // TODO (gkalpak): Remove once #137 is fixed.
        .concat(Observable.of(false)).take(1)
        .do(v => this.log(`Update available: ${v}`))
        .subscribe(v => v && this.activateUpdate());
  }

  private log(message: string) {
    const timestamp = (new Date).toISOString();
    this.logger.log(`[SwUpdates - ${timestamp}]: ${message}`);
  }
}
```

IgorMinar commented:
@gkalpak this looks good to me, except for the e2e test that seems to be legitimately failing.

mary-poppins commented:
You can preview b37aa08 at https://pr17699-b37aa08.ngbuilds.io/.

gkalpak commented:
Rebased on master and fixed the failing e2e test.

##
@petebacondarwin, I refactored `updateActivated` as you suggested, as it looks cleaner. I didn't switch to `Observable.timer`, because (a) it doesn't simplify things a lot (and I would have to change several tests) and (b) the current implementation is technically more correct (since we wait for the previous action to complete before initiating a new check). It won't change anything with our current 6-hour interval, but theoretically it is better 😛 (and it did make a difference when I was testing with 6 second intervals).

##
@IgorMinar, the reason that the e2e was failing is because the `aio-code` component imports and uses `MdSnackBar`, but the `MdSnackBarModule` was not imported. (This didn't fail before, because the `SwUpdates` service imported the `MdSnackBarModule`, which is no longer the case.)

What I find unexpected though is that there was no error when building. Shouldn't `yarn build` complain that it doesn't know where to get `MdSnackBar` from? Is it because `<aio-code>` is used in `entryComponents`?

mary-poppins commented:
You can preview 4c2afce at https://pr17699-4c2afce.ngbuilds.io/.
